sklearn/decomposition/_nmf.py

```diff
@@ -1304,24 +1246,97 @@ def fit_transform(self, X, y=None, W=None, H=None):
+        init = self.init
+        solver = self.solver
+        beta_loss = self.beta_loss
+        tol = self.tol
+        max_iter = self.max_iter
+        random_state = self.random_state
+        alpha = self.alpha
+        l1_ratio = self.l1_ratio
+        verbose = self.verbose
+        shuffle = self.shuffle
+        regularization = self.regularization
+        update_H = self.update_H
+
+        X = check_array(X, accept_sparse=('csr', 'csc'),
+                        dtype=[np.float64, np.float32])
+        check_non_negative(X, "NMF (input X)")
+        beta_loss = _check_string_param(solver, regularization,
+                                        beta_loss, init)
+
+        if X.min() == 0 and beta_loss <= 0:
+            raise ValueError("When beta_loss <= 0 and X contains zeros, "
+                             "the solver may diverge. Please add small values "
+                             "to X, or use a positive beta_loss.")
+
+        n_samples, n_features = X.shape
+        if n_components is None:
+            n_components = n_features
+
+        if not isinstance(n_components, numbers.Integral) or n_components <= 0:
+            raise ValueError("Number of components must be a positive integer;"
+                             " got (n_components=%r)" % n_components)
+        if not isinstance(max_iter, numbers.Integral) or max_iter < 0:
+            raise ValueError("Maximum number of iterations must be a positive "
+                             "integer; got (max_iter=%r)" % max_iter)
+        if not isinstance(tol, numbers.Number) or tol < 0:
+            raise ValueError("Tolerance for stopping criteria must be "
+                             "positive; got (tol=%r)" % tol)
+
+        # check W and H, or initialize them
+        if init == 'custom' and update_H:
+            _check_init(H, (n_components, n_features), "NMF (input H)")
+            _check_init(W, (n_samples, n_components), "NMF (input W)")
+            if H.dtype != X.dtype or W.dtype != X.dtype:
+                raise TypeError("H and W should have the same dtype as X. Got "
+                                "H.dtype = {} and W.dtype = {}."
+                                .format(H.dtype, W.dtype))
+        elif not update_H:
+            print(H)
+            _check_init(H, (n_components, n_features), "NMF (input H)")
+            if H.dtype != X.dtype:
+                raise TypeError("H should have the same dtype as X. Got "
+                                "H.dtype = {}.".format(H.dtype))
+            # 'mu' solver should not be initialized by zeros
+            if solver == 'mu':
+                avg = np.sqrt(X.mean() / n_components)
+                W = np.full((n_samples, n_components), avg, dtype=X.dtype)
+            else:
+                W = np.zeros((n_samples, n_components), dtype=X.dtype)
+        else:
+            W, H = _initialize_nmf(X, n_components, init=init,
+                                   random_state=random_state)
```

Comment:
I would move this in a dedicated private method, ``_initialize_W_H`` for the same reasons.
Or maybe split it into ```_initialize_W`` and ``_initialize_H`` such that we can call both in fit and only the first one in transform.